observation/store.go

```diff
@@ -53,76 +69,97 @@ type Result struct {
+	}
 
-			if neo4jErrorCode(err) == constraintError {
+	return results, nil
+}
 
-				log.Debug("constraint error identified - skipping the inserts for this instance",
-					log.Data{"instance_id": instanceID})
+func (store *Store) save(attempt, maxAttempts int, conn bolt.Conn, instanceID string, observations []*Observation) error {
+	query := buildInsertObservationQuery(instanceID, observations)
 
-				continue
-			}
+	dimensionIds, err := store.dimensionIDCache.GetNodeIDs(instanceID)
+	if err != nil {
+		store.reportError(instanceID, "failed to get dimension node id's", err)
+		return err
+	}
 
-			for instanceID := range instanceObservations {
-				store.reportError(instanceID, "observation batch insert failed", err)
-			}
+	queryParameters, err := createParams(observations, dimensionIds)
+	if err != nil {
+		store.reportError(instanceID, "failed to create query parameters for batch query", err)
+		return err
+	}
 
-			// todo: add retry logic and identify fatal errors that should be returned.
-			// any error will currently be sent to the reporter and mark the import as failed
-			// we do not return the error as we want the message to be consumed.
-			// returning an error causes the service to shutdown and not consume the message.
-			continue
-		}
+	queryResult, err := conn.ExecNeo(query, queryParameters)
+	if err != nil {
 
-		rowsAffected, err := queryResult.RowsAffected()
-		if err != nil {
-			store.reportError(instanceID, "error attempting to get number of rows affected in query result", err)
-			continue
+		if neo4jErrorCode(err) == constraintError {
+
+			log.Debug("constraint error identified - skipping the inserts for this instance",
+				log.Data{"instance_id": instanceID})
+
+			return err
 		}
 
-		log.Debug("save result",
-			log.Data{"rows affected": rowsAffected, "metadata": queryResult.Metadata()})
+		time.Sleep(getSleepTime(attempt, 20*time.Millisecond))
 
-		observationsInserted := int32(len(instanceObservations[instanceID]))
+		if attempt >= maxAttempts {
+			store.reportError(instanceID, "observation batch insert failed", ErrAttemptsExceededLimit{err})
 
-		result := &Result{
-			InstanceID:           instanceID,
-			ObservationsInserted: observationsInserted,
+			return ErrAttemptsExceededLimit{err}
 		}
 
-		results = append(results, result)
+		log.Info("got an error when saving observations, attempting to retry", log.Data{
+			"instance_id":  instanceID,
+			"retry_number": attempt,
+			"max_attempts": maxAttempts,
+		})
+
+		// Retry until it is successful or runs out of retries. TODO: the ability to retry
+		// neo queries should be built into go-ns, similar to the rchttp package.
+		return store.save(attempt+1, maxAttempts, conn, instanceID, observations)
 	}
 
-	return results, nil
+	rowsAffected, err := queryResult.RowsAffected()
+	if err != nil {
+		store.reportError(instanceID, "error attempting to get number of rows affected in query result", err)
+		return err
+	}
+
+	log.Debug("save result",
```

Comment:
I'm not sure this is the most descriptive log message (i realize you've only relocated it but a good chance to clear it up) - throughout this function the logs and error reports use conflicting language, it might help to standardize the terms used if possible? though anything more descriptive than `save result` would satisfy me :)